src/events/upgrade.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Event handler for handling OpensearchDashboards in-place upgrades."""
+import logging
+import time
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.upgrade import (
+    BaseModel,  # pyright: ignore [reportGeneralTypeIssues, reportAttributeAccessIssue]
+)
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    UpgradeGrantedEvent,
+)
+from typing_extensions import override
+
+if TYPE_CHECKING:
+    from charm import (
+        OpensearchDashboardsCharm,  # pyright: ignore [reportGeneralTypeIssues, reportAttributeAccessIssue]
+    )
+
+logger = logging.getLogger(__name__)
+
+
+class OpensearchDashboardsDependencyModel(BaseModel):
+    """Model for OpensearchDashboards Operator dependencies."""
+
+    osd_upstream: DependencyModel
+
+
+class ODUpgradeEvents(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "OpensearchDashboardsCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+
+    def post_upgrade_check(self) -> None:
+        """Runs necessary checks validating the unit is in a healthy state after upgrade."""
+        self.pre_upgrade_check()
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        pass
+
+    @override
+    def build_upgrade_stack(self) -> list[int]:
+        upgrade_stack = []
+        units = [self.charm.unit] + list(self.charm.state.peer_relation.units)
+        for unit in units:
+            upgrade_stack.append(int(unit.name.split("/")[-1]))
+
+        return upgrade_stack
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        logger.critical(
+            "\n".join(
+                [
+                    "Unit failed to upgrade and requires manual rollback to previous stable version.",
+                    "    1. Re-run `pre-upgrade-check` action on the leader unit to enter 'recovery' state",
+                    "    2. Run `juju refresh` to the previously deployed charm revision",
+                ]
+            )
+        )
+        return
+
+    @override
+    def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
+        self.charm.workload.stop()
+
+        if not self.charm.workload.install():
+            logger.error("Unable to install OpensearchDashboards...")
+            self.set_unit_failed()
```

Comment:
This seems to set an empty `BlockedStatus` on the unit, we should probably pass a `cause`